js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,416 @@
+					Metric: metrics.WSPing,
+					Time:   end,
+					Tags:   tags,
+					Value:  stats.D(time.Duration(avgPingRoundtrip)),
+				})
+			}
+
+			state.Samples = append(state.Samples, samples...)
+
+			return response
+		}
+	}
+}
+
+func (s *Socket) On(event string, handler goja.Value) {
+	if handler, ok := goja.AssertFunction(handler); ok {
+		s.eventHandlers[event] = append(s.eventHandlers[event], handler)
+	}
+}
+
+func (s *Socket) handleEvent(event string, args ...goja.Value) {
+	if handlers, ok := s.eventHandlers[event]; ok {
+		for _, handler := range handlers {
+			handler(goja.Undefined(), args...)
+		}
+	}
+}
+
+func (s *Socket) Send(message string) {
+	// NOTE: No binary message support for the time being since goja doesn't
+	// support typed arrays.
+	rt := common.GetRuntime(s.ctx)
+
+	writeData := []byte(message)
+	if err := s.conn.WriteMessage(websocket.TextMessage, writeData); err != nil {
+		s.handleEvent("error", rt.ToValue(err))
+	}
+
+	s.msgsSent++
+}
+
+func (s *Socket) Ping() {
+	rt := common.GetRuntime(s.ctx)
+	deadline := time.Now().Add(writeWait)
+	pingID := strconv.Itoa(s.pingSendCounter)
+	data := []byte(pingID)
+
+	err := s.conn.WriteControl(websocket.PingMessage, data, deadline)
+	if err != nil {
+		s.handleEvent("error", rt.ToValue(err))
+		return
+	}
+
+	s.pingSendTimestamps[pingID] = time.Now()
+	s.pingSendCounter++
+}
+
+func (s *Socket) trackPong(pingID string) {
+	pongTimestamp := time.Now()
+
+	if _, ok := s.pingSendTimestamps[pingID]; !ok {
+		// We received a pong for a ping we didn't send; ignore
+		// (this shouldn't happen with a compliant server)
+		return
+	}
+	pingTimestamp := s.pingSendTimestamps[pingID]
+
+	s.totalPingDuration += pongTimestamp.Sub(pingTimestamp)
+	s.totalPingCount++
+}
+
+func (s *Socket) SetTimeout(fn goja.Callable, timeoutMs int) {
+	// Starts a goroutine, blocks once on the timeout and pushes the callable
+	// back to the main loop through the scheduled channel
+	go func() {
+		select {
+		case <-time.After(time.Duration(timeoutMs) * time.Millisecond):
+			s.scheduled <- fn
+
+		case <-s.done:
+			return
+		}
+	}()
+}
+
+func (s *Socket) SetInterval(fn goja.Callable, intervalMs int) {
+	// Starts a goroutine, blocks forever on the ticker and pushes the callable
+	// back to the main loop through the scheduled channel
+	go func() {
+		ticker := time.NewTicker(time.Duration(intervalMs) * time.Millisecond)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.scheduled <- fn
+
+			case <-s.done:
+				return
+			}
+		}
+	}()
+}
+
+func (s *Socket) Close(args ...goja.Value) {
+	code := websocket.CloseGoingAway
+	if len(args) > 0 {
+		code = int(args[0].ToInteger())
+	}
+
+	s.closeConnection(code)
+}
+
+// Attempts to close the websocket gracefully
+func (s *Socket) closeConnection(code int) error {
+
+	var err error
+	s.shutdownOnce.Do(func() {
+		rt := common.GetRuntime(s.ctx)
+
+		err := s.conn.WriteControl(websocket.CloseMessage,
```

Comment:
You're using `:=` here, unintentionally shadowing the outer `err` var. This function will always return nil.

(This is why I always use another name instead of assigning directly to an outer variable like this ^^)